docs/evaluation/tutorials/eval_testing_tutorial.mdx

```diff
@@ -0,0 +1,813 @@
+    
+    testing.log_reference_outputs({"answer": 53})
+    
+    # Grab all the tool calls made by the LLM
+    tool_calls = [tc['function']['name'] for m in result['messages'] if m.additional_kwargs and 'tool_calls' in m.additional_kwargs for tc in m.additional_kwargs['tool_calls'] ]
+    
+    # This will log the number of steps taken by the LLM, which we can track over time to measure performance
+    testing.log_feedback(key="num_steps", value=len(result['messages']) - 1)
+    
+    # Assert that the code tool was used
+    assert "code_tool" in tool_calls
+    
+    # Assert that the answer is within 1 of the expected value
+    assert abs(result['structured_response']['numeric_answer'] - 53) <= 1`,
+    },
+    {
+      value: "vite",
+      label: "Vitest",
+      language: "typescript",
+      caption: `Remember to place each individual test inside a describe block \n
+      \`ls.describe("Agent Tests", () => {
+        // PLACE TESTS Here
+      })\`
+      `,
+      content: `ls.test(
+    "should execute code when needed",
+    {
+      inputs: { query: "What was the average return rate for FAANG stock in 2024?" },
+      expected: { answer: 53 },
+    },
+    async ({ inputs: { query }, expected: { answer } }) => {
+      const result = await agent.invoke({
+        messages: [["user", query]],
+      });
+      
+      ls.logOutputs(result);
+      
+      // Grab all the tool calls made by the LLM
+      const toolCalls = result.messages
+        .filter(m => m.additional_kwargs?.tool_calls)
+        .flatMap(m => m.additional_kwargs.tool_calls.map(tc => tc.function.name));
+
+      // This will log the number of steps taken by the LLM, which we can track over time to measure performance
+      ls.logFeedback({
+        key: "num_steps",
+        score: result.messages.length - 1, // The first message is the user message
+      });
+
+      // Assert that the tool calls include the "code_tool" function
+      expect(toolCalls).toContain("code_tool");
+
+      // Assert that the answer is within 1 of the expected answer
+      expect(Math.abs(result.structured_response.numeric_answer - answer)).toBeLessThanOrEqual(1);
+    }
+  );`,
+    },
+    {
+      value: "jest",
+      label: "Jest",
+      language: "typescript",
+      caption: `Remember to place each individual test inside a describe block \n
+      \`ls.describe("Agent Tests", () => {
+        // PLACE TESTS Here
+      })\`
+      `,
+      content: `ls.test(
+    "should execute code when needed",
+    {
+      inputs: { query: "What was the average return rate for FAANG stock in 2024?" },
+      expected: { answer: 53 },
+    },
+    async ({ inputs: { query }, expected: { answer } }) => {
+      const result = await agent.invoke({
+        messages: [["user", query]],
+      });
+      
+      ls.logOutputs(result);
+
+      // Grab all the tool calls made by the LLM
+      const toolCalls = result.messages
+        .filter(m => m.additional_kwargs?.tool_calls)
```

Comment:
For LangChain you can just use `m.tool_calls`

`additional_kwargs` is not necessary